File: src/client/controller/subpages/HomeController.java
package client.controller.subpages;

import client.controller.ClientApplicationController;
import client.model.BoggleApp.userInfo;
import client.model.subpages.HomeModel;
import client.model.subpages.LobbyModel;
import client.view.subpages.AvatarSelectionView;
import client.view.subpages.HomeView;
import client.view.subpages.SettingsView;
import server.model.DataPB;
import shared.SwingResources;

import javax.swing.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;

/**
 * The home controller processes user requests in joining the game, changing player profile picture, viewing
 * the global leaderboard and viewing the tutorial.
 */
public class HomeController {
    /**
     * The parent controller.
     */
    private ClientApplicationController parent;
    /**
     * The home view.
     */
    private HomeView homeView;
    /**
     * The model.
     */
    private HomeModel model;
    /**
     * The avatar selection view.
     */
    private AvatarSelectionView avatarSelectionView;
    /**
     * The settings view.
     */
    private SettingsView settingsView;

    /**
     * Constructs a HomeController with a specified view and model.
     * @param settingsView
     * @param homeView
     * @param model
     * @param parent
     */
    public HomeController(SettingsView settingsView, HomeView homeView, HomeModel model, ClientApplicationController parent) {
        this.settingsView = settingsView;
        this.homeView = homeView;
        this.model = model;
        this.parent = parent;

        parent.playDefaultMusic();

        // action listeners
        homeView.setJoinListener(new JoinGameListener(parent));
        homeView.setEditListener(new EditPfpListener());
        homeView.setTutorialListener(e -> parent.getView().showTutorial());

        homeView.getLblUsername().setText(model.getUsername());
        homeView.setAvatarImagePath(model.getPFPOFUser(model.getUsername()));

        // mouse listeners
        homeView.getBtnJoinGame().addMouseListener(new SwingResources.CursorChanger(homeView.getBtnJoinGame()));
        homeView.getBtnEditPfp().addMouseListener(new SwingResources.CursorChanger(homeView.getBtnEditPfp()));
        homeView.getBtnTutorial().addMouseListener(new SwingResources.CursorChanger(homeView.getBtnTutorial()));

         populateLeaderboard();
    }

    public void populateLeaderboard() {
        userInfo[] leaderboards = model.getLeaderboard();
        homeView.clearPnlLeaderBoard();
        for (userInfo leaderboard : leaderboards) {
            homeView.addPlayerInLeaderboard(leaderboard.username, leaderboard.pfpAddress, leaderboard.points);
        }
        homeView.revalidate();
        homeView.repaint();
    }

    /**
     * Processes the joining of a lobby.
     */
    static class JoinGameListener implements ActionListener {
        /**
         * The parent controller.
         */
        private ClientApplicationController parent;

        /**
         * Constructs a Join
         * @param parent
         */
        public JoinGameListener(ClientApplicationController parent) {
            this.parent = parent;
        }

        @Override
        public void actionPerformed(ActionEvent e) {
            SwingUtilities.invokeLater(() -> {
                parent.getView().showLobby();
                parent.getView().setNavLocationText("Lobby");System.out.println("creating new lobby");

                parent.getView().hideButtons();

                parent.playLobbyMusic();
                new LobbyController(new LobbyModel(parent.getModel().getUsername(), parent.getModel().getWfImpl()),
                        parent.getView().getLobbyView(), parent);
                System.out.println("new lobby created\n");

                /*
                parent.getView().getLobbyView().setExitLobbyListener(new ActionListener() {
                    @Override
                    public void actionPerformed(ActionEvent e) {
                      parent.getModel().getWfImpl().exitLobby(model.getUsername());
                        parent.getView().showHome();
                        parent.getView().setNavLocationText("Home");
                        parent.getView().showButtons();
                        parent.stopMusic();
                        parent.playDefaultMusic();
                    }
                });

                 */
            });
        }
    }


    /**
     * Processes editing of player profile picture.
     */
    class EditPfpListener implements ActionListener {
        @Override
        public void actionPerformed(ActionEvent e) {
            avatarSelectionView = new AvatarSelectionView(model.getUsername(), new DataPB(), homeView, settingsView);
        }
    }
}

File: src/Client_Python/yuen_client/Controller/subpages/HomeController.py
'''
from client.controller import ClientApplicationController
from client.model.BoggleApp import userInfo
from client.model.subpages import HomeModel,LobbyModel
from client.view.subpages import HomeView
import shared.SwingResources as swing_resources

import javax.swing as swing
import java.awt.event as event

class HomeController:
    def __init__(self, home_view, home_model, parent):
        self.view = view
        self.model = home_model
        self.parent = parent

        self.parent.play_default_music()

        # action listeners
        self.view.set_join_listener(self.JoinGameListener(self.parent))
        self.view.set_edit_listener(self.EditPfpListener())
        self.view.set_tutorial_listener(lambda: self.parent.get_view().show_tutorial())

        self.view.get_lbl_username().set_text(self.model.get_username())

        # mouse listeners
        self.view.get_btn_join_game().add_mouse_listener(SwingResources.CursorChanger(self. view.get_btn_join_game()))
        self.view.get_btn_edit_pfp().add_mouse_listener(SwingResources.CursorChanger(self.view.get_btn_edit_pfp()))
        self.view.get_btn_tutorial().add_mouse_listener(SwingResources.CursorChanger(self.view.get_btn_tutorial()))

        #populateLeaderboard is to be debugged in the future

    def populate_leaderboard(self):
        leaderboards = self.model.get_leaderboard()
        self.view.clear_pnl_leader_board()
        for leaderboard in leaderboards:
            self.view.add_player_in_leaderboard(leaderboard.username, leaderboard.pfp_address, leaderboard.points)
        self.view.revalidate()
        self.view.repaint()

    class JoinGameListener:
        def __init__(self, parent):
            self.parent = parent

        def __call__(self):
            def action():
                self.parent.get_view().show_lobby()
                self.parent.get_view().set_nav_location_text("Lobby")
                print("creating new lobby")

                self.parent.play_lobby_music()
                LobbyController(LobbyModel(self.parent.get_model().get_username(), self.parent.get_model().get_wf_impl()),self.parent.get_view().get_lobby_view(), self.parent)
                print("new lobby created\n")

                self.parent.get_view().get_lobby_view().set_exit_lobby_listener(lambda e: self.exit_lobby())

            import threading
            threading.Thread(target=action).start()

        def exit_lobby(self):
            # self.parent.get_model().get_wf_impl().exit_game_room(self.model.get_username())
            self.parent.get_view().show_home()
            self.parent.get_view().set_nav_location_text("Home")
            self.parent.stop_music()
            self.parent.play_default_music()

    class EditPfpListener:
        def __call__(self, event):
            pass
'''


